### ISSUE TITLE
Handling submitting tool output back to the agent

### ISSUE THREAD

**Issue by lynchian** [homanp/superagent#880]
### ⚠️ Please check that this feature request hasn't been suggested before.

- [X] I searched previous [Ideas in Discussions](https://github.com/homanp/superagent/discussions/categories/ideas) didn't find any similar feature requests.
- [X] I searched previous [Issues](https://github.com/homanp/superagent/labels/enhancement) didn't find any similar feature requests.

### 🔖 Feature description

- When a function is called by an agent, the response needs to be submitted to the agent in another invocation

- This creates problems in a workflow which automatically executes the invocation of the next agent as the next step

- Handling function calls to submit the tool output back to the agent without needing another invocation will solve the issue 

### Acknowledgements

- [X] My issue title is concise and descriptive.
- [X] I have searched the existing issues to make sure this feature has not been requested yet.
- [X] I have provided enough information for the maintainers to understand and evaluate this request.

**Comment by joaomirandas** [comment]
Yeah! I also want to see this in SuperAgent, execute actions is amazing but information should flow back to prompt to be use. If i'm not confuse, it's possible using workflows, but i think should be possible just defining tool and return value from there ( in terms of javascript/typescript )

**Comment by homanp** [comment]
> Yeah! I also want to see this in SuperAgent, execute actions is amazing but information should flow back to prompt to be use. If i'm not confuse, it's possible using workflows, but i think should be possible just defining tool and return value from there ( in terms of javascript/typescript )

I agree with you. I will add a way for this.

**Comment by joaomirandas** [comment]
I was wondering about how to do it and i reach 2 thoughts:

1) The way of "Custom Tools" works nowdays **it's great, and i don't think that should be removed**. It's very usefull in terms of API first service, I didn't see this kind of implementation allowing developer to use "tools" on your own code anywhere else. That's good and should not be removed;

2) We should implement additional something like "Actions" ( _just giving another cool name LoL_ ) which actually is an openAPI spec ( [_similar how Dify does_](https://github.com/langgenius/dify/) ) and this will capable developers using Superagent to call endpoints specified in that OpenAPI Specs and receive external data. I think about this because that operation could run inside **Superagent backend*, this will enable to use information received from external api as data inside methods and use that data to generate an rich output;

**Tips about it:**
I already tested dify implementation and it's quite good, but for some pieces sounds like beginner yet, we could have here some like some pattern to receive response from external API's and i also suggest to always require an field with "message" to give generative context about how that request perform. ( eg. if request about get wether succeed should return in response.message "weather acquisition is ok and data is available" - i miss that kind of response context in dify implementation and that make output generation not good enouth. )

Check that part of Dify, it's most similar that i could find.

![Image](https://github.com/homanp/superagent/assets/29043431/2e249eea-4f24-4b10-8357-a3b06805d04d)

What did you think about this approach @homanp ? and you @lynchian ?

**Comment by homanp** [comment]
> I was wondering about how to do it and i reach 2 thoughts:
> 
> 1. The way of "Custom Tools" works nowdays **it's great, and i don't think that should be removed**. It's very usefull in terms of API first service, I didn't see this kind of implementation allowing developer to use "tools" on your own code anywhere else. That's good and should not be removed;
> 2. We should implement additional something like "Actions" ( _just giving another cool name LoL_ ) which actually is an openAPI spec ( [_similar how Dify does_](https://github.com/langgenius/dify/) ) and this will capable developers using Superagent to call endpoints specified in that OpenAPI Specs and receive external data. I think about this because that operation could run inside *_Superagent backend_, this will enable to use information received from external api as data inside methods and use that data to generate an rich output;
> 
> **Tips about it:** I already tested dify implementation and it's quite good, but for some pieces sounds like beginner yet, we could have here some like some pattern to receive response from external API's and i also suggest to always require an field with "message" to give generative context about how that request perform. ( eg. if request about get wether succeed should return in response.message "weather acquisition is ok and data is available" - i miss that kind of response context in dify implementation and that make output generation not good enouth. )
> 
> Check that part of Dify, it's most similar that i could find.
> 
> ![Image](https://private-user-images.githubusercontent.com/29043431/310749633-2e249eea-4f24-4b10-8357-a3b06805d04d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk3OTA4MTIsIm5iZiI6MTcwOTc5MDUxMiwicGF0aCI6Ii8yOTA0MzQzMS8zMTA3NDk2MzMtMmUyNDllZWEtNGYyNC00YjEwLTgzNTctYTNiMDY4MDVkMDRkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzA3VDA1NDgzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEwMDI5MDQ2MmVlMTg5NjdkZDk5NTcyN2ZmMTQ3ZjE1NDc1N2FkYzVjNmNiYmQxOWJhNDI1YTZhNTliMjljOWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.dcYSeW_yCQgzYAPS8if7829ubX9C9Dk3e41J6AAau_U)
> 
> What did you think about this approach @homanp ? and you @lynchian ?

So you wan to add openapi spec tool calling? Sure we can do that. Would you be willing to make a contribution here? Also that wouldn't really solve the problem for this usecase. But I have a solution for that.

**Comment by lynchian** [comment]
For a direct example of implementation, this is how OpenAI Assistants do it :

https://github.com/openai/openai-python/blob/e41abf7b7dbc1e744d167f748e55d4dedfc0dca7/src/openai/resources/beta/threads/runs/runs.py#L314

**Comment by homanp** [comment]
@elisalimli let's implement this.

**Comment by elisalimli** [comment]
@lynchian @joaomirandas 

We don't think this would really work with workflows that has multiple agents.

Let's say a workflow has:

- Agent 1
- Agent 2 (this agent has a custom function calling)
- Agent 3

If we run this workflow, and agent 2 decides to call the given custom function, what we will do here? We can not move on to the next agent or stop responding. Because if we stop, we will lost the agent 1's response. 

Nevertheless, we do want to support this feature. But for /agents endpoint only.

@lynchian @joaomirandas What are your takes on this?
